server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4950,6 +4988,260 @@ public void testKeepMinRetainedSeqNoByMergePolicy() throws IOException {
+            engine.delete(replicaDeleteForDoc("1", 5, 4, threadPool.relativeTimeInMillis()));
+            List<Tuple<Long, Long>> seqNos = readUpdatedBySeqNos.apply(engine);
+            // index#0 -> index#1 -> delete#3 -> index#2 -> index#5 -> delete#4
+            assertThat(seqNos, containsInAnyOrder(
+                Tuple.tuple(0L, 1L),   // 0 -> 1 - seq#0 should be updated once because of refresh
+                Tuple.tuple(1L, 3L),   // 1 -> 5 (deleted by #3, then updated by #5)
+                Tuple.tuple(2L, 3L),   // index#2 is stale by delete#3, thus should never be updated again.
+                Tuple.tuple(3L, null), // the first delete tombstone without updated_by_seqno
+                Tuple.tuple(3L, 5L),   // the second delete tombstone without updated_by_seqno
+                Tuple.tuple(4L, 5L),   // delete#4 is stale, thus having updated_by_seqno=5
+                Tuple.tuple(5L, null)  // still alive
+            ));
+        }
+        // On primary
+        try (Store store = createStore();
+             InternalEngine engine = createEngine(config(indexSettings, store, createTempDir(),
+                 newMergePolicy(), null, null, globalCheckpoint::get))) {
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.refresh("test");
+            engine.delete(new Engine.Delete("test", "1", newUid("1"), primaryTerm.get()));
+            engine.index(indexForDoc(createParsedDoc("1", null)));
+            engine.delete(new Engine.Delete("test", "1", newUid("1"), primaryTerm.get()));
+            engine.refresh("test");
+            List<Tuple<Long, Long>> seqNos = readUpdatedBySeqNos.apply(engine);
+            // index#0 -> index#1 -> delete#2 -> index#3 -> delete#4
+            assertThat(seqNos, containsInAnyOrder(
+                Tuple.tuple(0L, 1L),   // index#0 -> index#1 -> refresh
+                Tuple.tuple(1L, 4L),   // index#1 -> delete#2 -> index#3 -> delete#4
+                Tuple.tuple(2L, null), // the first delete tombstone without updated_by_seqno
+                Tuple.tuple(2L, 3L),   // the second delete tombstone with updated_by_seqno
+                Tuple.tuple(3L, 4L),   // index#3 -> delete#4
+                Tuple.tuple(4L, null)  // still alive
+            ));
+        }
+    }
+
+    @TestLogging("_ROOT:DEBUG")
+    public void testKeepDocsForRollback() throws Exception {
+        IOUtils.close(engine, store);
+        threadPool = spy(threadPool);
+        AtomicLong clockTime = new AtomicLong();
+        when(threadPool.relativeTimeInMillis()).thenAnswer(i -> clockTime.incrementAndGet());
+        List<Engine.Operation> operations = new ArrayList<>();
+        int numOps = scaledRandomIntBetween(10, 200);
+        for (int seqNo = 0; seqNo < numOps; seqNo++) {
+            String id = Integer.toString(between(1, 5));
```

Comment:
why not use the random history generator? maybe we should extend it if we need to?

Reply:
I adjusted the history generator method and use it here.